packages/compass-sidebar/src/components/use-filtered-connections.ts

```diff
@@ -341,20 +339,63 @@ type UseFilteredConnectionsHookResult = {
     databaseId: string,
     isExpanded: boolean
   ): void;
-  onConnectionsChanged(this: void, newConnections: ConnectionInfo[]): void;
 };
 
-export const useFilteredConnections = (
-  connections: SidebarConnection[],
-  filterRegex: RegExp | null,
-  _fetchAllCollections: () => void,
-  _onDatabaseExpand: (connectionId: string, databaseId: string) => void
-): UseFilteredConnectionsHookResult => {
+function filteredConnectionsToSidebarConnection(
+  filteredConnections: FilteredConnection[]
+): SidebarConnection[] {
+  const sidebarConnections: SidebarConnection[] = [];
+  for (const connection of filteredConnections) {
+    if (connection.connectionStatus === ConnectionStatus.Connected) {
+      sidebarConnections.push({
+        ..._.omit(connection, ['isMatch']),
+        databases: connection.databases.map((database) => {
+          return {
+            ..._.omit(database, ['isMatch']),
+            collections: database.collections.map((collection) =>
+              _.omit(collection, ['isMatch'])
+            ),
+          };
+        }),
+      });
+    } else {
+      sidebarConnections.push({
+        ..._.omit(connection, ['isMatch']),
+      });
+    }
+  }
+  return sidebarConnections;
+}
+
+export const useFilteredConnections = ({
+  connections,
+  filterRegex,
+  fetchAllCollections,
+  onDatabaseExpand,
+}: {
+  connections: SidebarConnection[];
+  filterRegex: RegExp | null;
+  fetchAllCollections: () => void;
+  onDatabaseExpand: (connectionId: string, databaseId: string) => void;
+}): UseFilteredConnectionsHookResult => {
   const [{ filtered, expanded }, dispatch] = useReducer(connectionsReducer, {
     filtered: undefined,
     expanded: {},
   });
 
+  const activeConnections = useMemo(() => {
+    return connections
+      .filter(({ connectionStatus }) => {
+        return connectionStatus === ConnectionStatus.Connected;
+      })
+      .map(({ connectionInfo }) => connectionInfo);
+  }, [connections]);
+
+  // get rid of stale connection related metadata in the state
+  useEffect(() => {
+    dispatch({ type: CONNECTIONS_CHANGED, connections: activeConnections });
```

Comment:
Shouldn't all the logic for determining what's active / expanded be in actions and reducers rather than in an effect inside a hook that triggers when the hook is used?

Reply:
True that, infact it used to be in reducer but at some point we moved it to a react state because the filtering was not just for databases but also for connections which were not in redux. Moving this back to redux would require some refactoring in the slice to also include connections related data in there.

The idea of this hook is to serve as an intermediary for further refactoring to follow which is planned in https://jira.mongodb.org/browse/COMPASS-7982